### ISSUE TITLE
Exception: An internal error occurred during: "calculating highlighting".

### ISSUE THREAD

**Issue by miklossy** [eclipse/xtext-eclipse#1389]
Just hit the following exception while implementing the Xtext Domainmodel Tutorial:

- Eclipse 2020-03
- Xtext 2.22.0.v20200404-0123

![image](https://user-images.githubusercontent.com/20393472/78460458-3f3dc580-76c1-11ea-93b1-bcfb61b3997c.png)

```
java.lang.RuntimeException: org.eclipse.xtext.service.OperationCanceledError: org.eclipse.core.runtime.OperationCanceledException
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:71)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:525)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:497)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:136)
	at org.eclipse.xtext.util.concurrent.IReadAccess.tryReadOnly(IReadAccess.java:50)
	at org.eclipse.xtext.util.concurrent.IReadAccess.tryReadOnly(IReadAccess.java:71)
	at org.eclipse.xtext.ui.editor.syntaxcoloring.HighlightingReconciler$2.run(HighlightingReconciler.java:352)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.eclipse.xtext.service.OperationCanceledError: org.eclipse.core.runtime.OperationCanceledException
	at org.eclipse.xtext.service.OperationCanceledManager.asWrappingOperationCanceledException(OperationCanceledManager.java:72)
	at org.eclipse.xtext.service.OperationCanceledManager.throwOperationCanceledException(OperationCanceledManager.java:78)
	at org.eclipse.xtext.service.OperationCanceledManager.checkCanceled(OperationCanceledManager.java:87)
	at org.eclipse.xtext.xbase.resource.BatchLinkableResource.resolveLazyCrossReferences(BatchLinkableResource.java:168)
	at org.eclipse.xtext.xbase.resource.BatchLinkableResource.linkBatched(BatchLinkableResource.java:196)
	at org.eclipse.xtext.ui.editor.syntaxcoloring.HighlightingReconciler.beforeRefresh(HighlightingReconciler.java:379)
	at org.eclipse.xtext.ui.editor.syntaxcoloring.HighlightingReconciler$2$1.exec(HighlightingReconciler.java:356)
	at org.eclipse.xtext.ui.editor.syntaxcoloring.HighlightingReconciler$2$1.exec(HighlightingReconciler.java:1)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:27)
	at org.eclipse.xtext.util.concurrent.WrappingCancelableUnitOfWork.exec(WrappingCancelableUnitOfWork.java:58)
	at org.eclipse.xtext.util.concurrent.CancelableUnitOfWork.exec(CancelableUnitOfWork.java:27)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:69)
	... 7 more
Caused by: org.eclipse.core.runtime.OperationCanceledException
	at org.eclipse.xtext.service.OperationCanceledManager.getPlatformSpecificOperationCanceledException(OperationCanceledManager.java:82)
	... 18 more
```

**Comment by miklossy** [comment]
Could be someting related to Xtend.

**Comment by cdietrich** [comment]
Can you reproduce this with 2.21 release?

**Comment by miklossy** [comment]
Not really. I saw that problem yesterday. Today I am working with the same environment for a while and have not seen that problem till now. It could be something that occurs sporadically.

@cdietrich Can you extract some information from the stracktrace that could help us further identify/reproduce the problem?

**Comment by miklossy** [comment]
It could be something to do with the refactoring from Xtend to Java, so that the exception thrown by the `work.exec(param)` call became visible now through this [code](https://github.com/eclipse/xtext-core/blob/9158b09e19a3264abb71cb425c18f87eb9da4773/org.eclipse.xtext/src/org/eclipse/xtext/resource/OutdatedStateManager.java#L71).

**Comment by cdietrich** [comment]
This is why I am asking. Was the exception catched and swallowed before?

**Comment by miklossy** [comment]
Yes, that's the point, see my [comments](https://github.com/eclipse/xtext-eclipse/issues/1389#issuecomment-609378686) above.

**Comment by cdietrich** [comment]
How was the old generated code?

**Comment by miklossy** [comment]
You can find the old generated code [here](https://github.com/eclipse/xtext-core/blob/8fca1c654650f785de8d762d00784103db53b189/org.eclipse.xtext/xtend-gen/org/eclipse/xtext/resource/OutdatedStateManager.java#L101-L103)

**Comment by cdietrich** [comment]
@szarnekow any idea?

**Comment by cdietrich** [comment]
So the old code had a different set of try and catch

**Comment by cdietrich** [comment]
@miklossy can you also check if thecalling code was migrated

**Comment by miklossy** [comment]
> 
> 
> So the old code had a different set of try and catch

Exactly.

**Comment by miklossy** [comment]
> 
> 
> @miklossy can you also check if thecalling code was migrated (HighlightingReconciler)

No. The [HighlightingReconciler](https://github.com/eclipse/xtext-eclipse/commits/master/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/syntaxcoloring/HighlightingReconciler.java) was originally implemented in Java.

**Comment by cdietrich** [comment]
Would still help if this was somehow reproducible

**Comment by cdietrich** [comment]
public static <T> T throwUncheckedException(Throwable e) {
		if (e instanceof RuntimeException)
			throw (RuntimeException) e;
		if (e instanceof Error)
			throw (Error) e;
		if (e instanceof Exception)
			throw new WrappedException((Exception) e);
		throw new RuntimeException(e);
	}

public final class OperationCanceledException extends RuntimeException {

=> before we rethrow. now we wrap
=> callers might see a different exception now

**Comment by cdietrich** [comment]
=> elmininating sneaky throw complely might not be a so good idea
cc @tivervac @szarnekow 

how shall be proceed. call the sneeky throw again?

**Comment by cdietrich** [comment]
we should also go through other PRs for that pattern

**Comment by miklossy** [comment]
> 
> 
> Would still help if this was somehow reproducible

I found something reproducible: If I execute the [XtextGrammarQuickfixProviderTests](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.xtext.ui.tests/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixProviderTest.java) test cases, I see the same exception in the console:

![screencast](https://user-images.githubusercontent.com/20393472/78471224-f5d88f00-772f-11ea-8229-0e66a329f55a.gif)


@cdietrich Do you see that also on your computer?  Can you find that stacktrace somewhere in the Jenkins log?

**Comment by cdietrich** [comment]
yes it get it

**Comment by cdietrich** [comment]
@miklossy see the log files in the results

**Comment by cdietrich** [comment]
(that seem to be not there for unknow reaons)

**Comment by cdietrich** [comment]
cause

    failure {
      archiveArtifacts artifacts: '**/target/work/data/.metadata/.log, **/hs_err_pid*.log'
    }

**Comment by cdietrich** [comment]
=> push a branch with a failing test and you will get all logs

**Comment by miklossy** [comment]
The problem is that the XtextGrammarQuickfixProviderTest test cases still passes, although exception is thrown.

**Comment by cdietrich** [comment]
test and arbritrary exceptions in log have nothing todo with each other in our setup

**Comment by szarnekow** [comment]
> => elmininating sneaky throw complely might not be a so good idea
> cc @tivervac @szarnekow
> 
> how shall be proceed. call the sneeky throw again?

I'd prefer a case-by-case decision.
Throwing checked exceptions without those being declared on the method signature does not allow a catch clause to be used for that specific exception.
Instead of sneakyThread, I'd prefer `throwUnchecked` which wraps only if necessary. We should even consider wrapping in a specific runtime exception similar to the idiom used for UncheckedExecutionException or RuntimeIOException.

**Comment by cdietrich** [comment]
Fixed in 2.22
